public/script.js

```diff
@@ -8487,99 +8513,143 @@ export function callPopup(text, type, inputValue = '', { okButton, rows, wide, w
+ * @returns {boolean}
  */
-export function refreshSwipeButtons() {
-    hideSwipeButtons();
-    showSwipeButtons();
+export function isSwipingAllowed() {
+    return (
+        //Swipe cannot be called on an empty chat.
+        chat.length !== 0 &&
+        //The swipes setting must be enabled, and swipes can't be hidden.
+        swipes && !swipesHidden &&
+        //Cannot swipe while generating.
+        !isGenerating() &&
+        //If mid-swipe, the message cannot be swiped.
+        swipeState === SWIPE_STATE.NONE
+    );
 }
 
-export function showSwipeButtons(mesId = chat.length - 1) {
-    isSwipingAllowed = true;
-
-    if (chat.length === 0) {
-        return;
-    }
+/**
+ * Returns true if the message is swipeable.
+ * This does not check if messages are generally swipeable. See isSwipingAllowed().
+ * This does not check if the swipes exist or are valid.
+ * @param {number} messageId The message Id to check.
+ * @param {object} [message=undefined] If undefined, then the message checks will be skipped.
+ * @returns {boolean}
+ */
+export function isMessageSwipeable(messageId, message = undefined) {
+    message ??= chat[messageId];
 
     if (
-        chat[mesId].is_system ||
-        !swipes ||
-        Number($('.mes:last').attr('mesid')) < 0 ||
-        chat[mesId].is_user ||
-        (selected_group && is_group_generating)
-    ) {
-        return;
-    }
-
-    // swipe_id should be set if alternate greetings are added
-    if (chat.length == 1 && chat[0].swipe_id === undefined) {
-        return;
-    }
-
-    //had to add this to make the swipe counter work
-    //(copied from the onclick functions for swipe buttons..
-    //don't know why the array isn't set for non-swipe messages in Generate or addOneMessage..)
-    if (chat[mesId]['swipe_id'] === undefined) {              // if there is no swipe-message in the last spot of the chat array
-        chat[mesId]['swipe_id'] = 0;                        // set it to id 0
-        chat[mesId]['swipes'] = [];                         // empty the array
-        chat[mesId]['swipes'][0] = chat[mesId]['mes'];  //assign swipe array with last message from chat
-        chat[mesId]['swipe_info'] = [];
-        chat[mesId]['swipe_info'][0] = {
-            'send_date': chat[mesId]['send_date'],
-            'gen_started': chat[mesId]['gen_started'],
-            'gen_finished': chat[mesId]['gen_finished'],
-            'extra': structuredClone(chat[mesId]['extra']),
-        };
-    }
+        //Only messages below the currently edited message can be swiped, if it's not mid-swipe edit.
+        ((messageId > (this_edit_mes_id ?? -1)) && (swipeState != SWIPE_STATE.EDITING)) &&
+
+        //If the message is the last message, and it exists.
+        (messageId == chat.length - 1) &&
+        (message &&
+            //Some messages, like the welcome screen, are not swipeable.
+            !(message?.['extra']?.['swipeable'] === false) &&
```

Comment:
Please add this property to `ChatMessageExtra` type

Reply:
Also you can use direct access instead of the indexer
```suggestion
            !(message?.extra?.swipeable === false) &&
```

Reply:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4712/commits/a981a157ba8af083959939977e16b9b16857ca36